File: OOPRecords.Model/TeacherRepository.cs
using NakedObjects;
using System.Linq;

namespace OOPRecords.Model
{
public class TeacherRepository
{

        public IDomainObjectContainer Container { set; protected get; }

        public IQueryable<Teacher> AllTeachers()
        {
            return Container.Instances<Teacher>();
        }

        public IQueryable<Teacher> FindTeacherByLastName(string lastName)
        {
            return from t in AllTeachers()
                   where t.LastName.ToUpper().Contains(lastName.ToUpper())
                   select t;
        }

        public IQueryable<Teacher> FindTeacherByJobTitle(string job)
        {
            return from t in AllTeachers()
                   where t.JobTitle.ToUpper().Contains(job.ToUpper())
                   select t;
        }

        public Teacher NewTeacher(string title, string lastName, string job)
        {
            var t = Container.NewTransientInstance<Teacher>();
            t.Title = title;
            t.LastName = lastName;
            t.JobTitle = job;
            Container.Persist(ref t);
            return t;
        }
    }
}

File: OOPRecords.ModelVB/TeacherRepository.vb


Imports NakedObjects

Public Class TeacherRepository
    Private _container As IDomainObjectContainer
    Public WriteOnly Property Container() As IDomainObjectContainer
        Set(ByVal value As IDomainObjectContainer)
            _container = value
        End Set
    End Property

    Public Function AllTeachers() As IQueryable(Of Teacher)
        Return _container.Instances(Of Teacher)()
    End Function

    Public Function FindTeacherByLastName(ByVal lastName As String) As IQueryable(Of Teacher)
        Return From t In AllTeachers() Where t.LastName.ToUpper().Contains(lastName.ToUpper()) Select t
    End Function

    Public Function FindTeacherByJobTitle(ByVal job As String) As IQueryable(Of Teacher)
        Return From t In AllTeachers() Where t.JobTitle.ToUpper().Contains(job.ToUpper()) Select t
    End Function

    Public Function NewTeacher(ByVal title As String, ByVal lastName As String, ByVal job As String) As Teacher
        Dim t = _container.NewTransientInstance(Of Teacher)()
        t.Title = title
        t.LastName = lastName
        t.JobTitle = job
        _container.Persist(t)
        Return t
    End Function
End Class
